SCClassLibrary/Common/Core/CondVar.sc

```diff
@@ -0,0 +1,105 @@
+CondVar {
+	var waitingThreads, timeoutThreads;
+
+	*new { ^super.newCopyArgs(Array.new, Array.new) }
+
+	wait { |predicate|
+		if(predicate.notNil) {
+			while { predicate.value.not } { this.wait }
+		} {
+			this.prSleepThread(thisThread);
+			\wait.yield
+		}
+	}
+
+	waitFor { |timeoutBeats, predicate|
+		var endingBeats, waitingThread, timeoutThread, enterBeats = thisThread.beats;
+		var didNotTimeout = true;
+
+		// May throw if preconditions are not met
+		timeoutBeats = this.prHandleTimeoutPreconditions(timeoutBeats);
+
+		if(timeoutBeats <= 0) { ^predicate.notNil and: { predicate.value } };
+
+		if(predicate.notNil) {
+			endingBeats = enterBeats + timeoutBeats;
+			// will exit when predicate is true or timeout has expired
+			while {
+				predicate.value.not
+			} {
+				if(this.waitFor(endingBeats - thisThread.beats).not) { ^predicate.value }
+			};
+
+			^true // we were signalled and predicate was true
+		} {
+			waitingThread = thisThread;
+			timeoutThread = Routine {
+				var wokenThread;
+				// TODO is this right? can we be more precise?
+				timeoutBeats.wait;
+
+				// wokenThread must be non-nil. If the thread woke before this timeout fired, this routine will already
+				// be stopped and we won't get here.
+				didNotTimeout = false;
+				wokenThread = this.prRemoveWaitingThread(waitingThread);
+				timeoutThreads.remove(timeoutThread);
+				this.prWakeThread(wokenThread);
+			};
+			timeoutThreads = timeoutThreads.add(timeoutThread);
+			timeoutThread.play(thisThread.clock);
+			this.wait;
+			// TODO will this be gc'd properly?
+			timeoutThreads.remove(timeoutThread).stop;
+
+			^didNotTimeout
+		}
+	}
+
+	signalOne {
+		if (waitingThreads.isEmpty.not) {
+			this.prWakeThread(waitingThreads.removeAt(0));
+		}
+	}
+
+	signalAll {
+		waitingThreads.do(this.prWakeThread(_));
+		waitingThreads = Array.new;
+	}
+
+	shallowCopy { this.shouldNotImplement(thisMethod) }
+	deepCopy { this.shouldNotImplement(thisMethod) }
+
+	// abstracted to ease experimenting with using `t` vs. `t.threadPlayer`
+
+	// non-thread player versions
+	/*
+	prSleepThread { |t| waitingThreads = waitingThreads.add(t) }
+	prWakeThread { |t| t.clock.sched(0, t) }
+	prRemoveWaitingThread { |t| ^waitingThreads.remove(t) }
+	*/
+
+	// thread player versions
+	prSleepThread { |t| waitingThreads = waitingThreads.add(t.threadPlayer) }
+	prWakeThread { |t| t.clock.sched(0, t) }
+	prRemoveWaitingThread { |t| ^waitingThreads.remove(t.threadPlayer) }
+
+	// Precondition checks for `waitFor` timeout. May throw.
+	prHandleTimeoutPreconditions { |n|
```

Comment:
"handle preconsitions" is a bit general for what it does. But I would suggest to introduce an interface for the respective classes, because it is generally useful and sematically unambiguous in all other use cases, I think. Then name `asTimeoutValue` could also be something more general like `asDeltaTime`? Here I assume for simplicity that an Object that responds to `asInteger` will also respond to `asFloat`.

```
+ SimpleNumber {
	asTimeoutValue {
		if(this.isNaN) { Error("timeout value should be a number").throw };
		if(this === inf ) { Error("timeout value should be finite").throw };
		if(this < 0 ) { Error("timeout value should be positive").throw };
		^this
	}
	
}

+ Object {
	
	asTimeoutValue {
		^this.asFloat.asTimeoutValue
	}
}

```

Then this method isn't needed, and above you'd write instead of:

`timeoutBeats = this.prHandleTimeoutPreconditions(timeoutBeats);`

`timeoutBeats = timeoutBeats.asTimeoutValue`

Reply:
I won't do that here, but you are welcome to try it out in a future PR if you find more use cases for this concept.

Reply:
at this point, I am really just looking for feedback on whether the restrictions on the type and value make sense. to me, they seem intuitive (except for maybe inf and -inf), and i think it's important to not allow an exception to propagate out of the spawned timeout thread, but this still feels like a rough edge in the design to me and I'd be open to other approaches

Reply:
Yes, this is a good procedure, you can do it this way. In the final version, these conversions should be replaced by something less particular (I don't remember any other case where we have something like this), but that can wait for a bit later if you prefer.